nx/source/services/hid.c

```diff
@@ -69,6 +106,135 @@ void* hidGetSharedmemAddr(void) {
     return shmemGetAddr(&g_hidSharedmem);
 }
 
+void hidSetControllerLayout(HIDControllerID id, HIDControllerLayoutType layoutType) {
+    controllerLayout[id] = layoutType;
+}
+
+void hidScanInput(void) {
```

Comment:
I think we might wanna either split hidScanInput into:

hidScanKeyboard()
hidScanController()
hidScanTouch()
hidScanMouse()
etc...

or have it take a bitfield like this:

hidScanInput(HidScan_Controller | HidScan_Touch)

Reply:
For performance reasons

Reply:
Or only process the currently activated devices?

Reply:
I think controller state (and touch?) should be unconditionally scanned. The other devices (keyboard, mouse) should only be scanned if they are active/enabled. I don't think `hidScanInput` should be split into multiple functions, or a parameter added to it.

Reply:
Why have a special rule? It's better to have the API symmetric for all types of input

Reply:
But not all types of input will be as commonly used. I think button input should always be available.

Reply:
Doesn't justify a special rule.

A function should not assume what the user wants to do. It should do what it says, and only that.

Reply:
Actually, is it possible to disable controller input in the first place? If so, then there wouldn't really be a way to unconditionally disable button input.

I think libnx could keep track of which sources of input are enabled at the moment, and scan those in `hidScanInput`, therefore obviating the need for a special parameter.

Reply:
I'm thinking what I'll do is add the activate/deactivate commands @yellows8 mentioned, keep keyboard+mouse off by default at init and then have libnx keep track of and only update button states for devices which are enabled.

I think I can also optimize controllers a bit by checking if the slot has no controller attached, HIDControllerType should have no bits set with controllers detached. I'm not sure if I have anything similar to detect mouse+keyboard though, I'll have to investigate a bit I suppose.